mula/scheduler/schedulers/scheduler.py

```diff
@@ -245,119 +248,140 @@ def push_item_to_queue(self, p_item: models.PrioritizedItem) -> None:
-    ) -> None:
-        """Make a function run in a thread, and add it to the dict of threads.
+            self.post_pop(p_item)
+
+        return p_item
+
+    def post_pop(self, p_item: models.Task) -> None:
+        """When a boefje task is being removed from the queue. We
+        persist a task to the datastore with the status RUNNING
 
         Args:
-            name: The name of the thread.
-            target: The function to run in the thread.
-            interval: The interval to run the function.
-            daemon: Whether the thread should be a daemon.
-            loop: Whether the thread should loop.
+            p_item: The prioritized item from the priority queue.
         """
-        t = utils.ThreadRunner(
-            name=name,
-            target=target,
-            stop_event=self.stop_event_threads,
-            interval=interval,
-            daemon=daemon,
-            loop=loop,
-        )
-        t.start()
+        # Update task
+        task = self.ctx.datastores.task_store.get_task(str(p_item.id))
+        if task is None:
+            self.logger.warning(
+                "PrioritizedItem %s popped from %s, task %s not found in datastore, could not update task status",
+                p_item.id,
+                self.queue.pq_id,
+                p_item.data.get("id"),
+                p_item_id=p_item.id,
+                task_id=p_item.data.get("id"),
+                queue_id=self.queue.pq_id,
+                scheduler_id=self.scheduler_id,
+            )
+            return
 
-        self.threads.append(t)
+        task.status = models.TaskStatus.DISPATCHED
+        self.ctx.datastores.task_store.update_task(task)
 
-    def is_space_on_queue(self) -> bool:
-        """Check if there is space on the queue.
+        self.last_activity = datetime.now(timezone.utc)
 
-        NOTE: maxsize 0 means unlimited
+    def signal_handler_task(self, task: models.Task) -> None:
+        """Handle a task that has been completed or failed."""
+        if task.status not in [models.TaskStatus.COMPLETED, models.TaskStatus.FAILED]:
+            return
 
-        Returns:
-            True if there is space on the queue, False otherwise.
+        def _calculate_deadline(task: models.Task):
+            schema = self.ctx.datastores.schema_store.get_schema_by_hash(task.hash)
+
+            try:
+                schema.deadline_at = datetime.fromtimestamp(self.deadline_ranker.rank(schema))
+            except Exception:
+                self.logger.error(
+                    "Unable to calculate deadline for schedule %s. Disabling schedule",
+                    schema.id,
+                    schema_id=schema.id,
+                    task_hash=task.p_item.hash,
+                    scheduler_id=self.scheduler_id,
+                )
+                schema.enabled = False
+
+            self.ctx.datastores.schema_store.update_schema(schema)
+
+        # TODO: explain why we're using executor here
```

Comment:
Calculation of deadlines could potentially take a long time. Since this is called from the server API, we want to have it be calculated in the background since we want to return a response without keeping the connection open.